src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs

```diff
@@ -84,205 +109,340 @@ public void Initialize()
-            {
-                this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(false, blockDisconnectedData.DisconnectedBlock));
-            });
+                        lastFlushTime = DateTime.Now;
 
-            this.nodeStats.RegisterStats(this.AddComponentStats, StatsType.Component, 400);
+                        this.logger.LogDebug("Flush completed.");
+                    }
 
-            if ((this.consensusManager.Tip.HashBlock.ToString() != this.addressesIndex.TipHash))
-            {
-                const string message = "TransactionIndexer is in inconsistent state. This can happen if you've enabled txindex on an already synced or partially synced node. " +
-                                       "Remove everything from the data folder and run the node with -txindex=true.";
+                    if (this.cancellation.IsCancellationRequested)
+                        break;
 
-                this.logger.LogCritical(message);
-                this.logger.LogTrace("(-)[INCONSISTENT_STATE]");
-                throw new Exception(message);
-            }
+                    ChainedHeader nextHeader = this.consensusManager.Tip.GetAncestor(this.IndexerTip.Height + 1);
 
-        }
+                    if (nextHeader == null)
+                    {
+                        this.logger.LogDebug("Next header wasn't found. Waiting.");
 
-        private Task OnEnqueueAsync(KeyValuePair<bool, ChainedHeaderBlock> item, CancellationToken cancellationtoken)
-        {
-            try
-            {
-                this.ProcessBlockAddedOrRemoved(item);
+                        try
+                        {
+                            await Task.Delay(DelayTimeMs, this.cancellation.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        continue;
+                    }
+
+                    if (nextHeader.Previous.HashBlock != this.IndexerTip.HashBlock)
+                    {
+                        ChainedHeader lastCommonHeader = nextHeader.FindFork(this.IndexerTip);
+
+                        this.logger.LogDebug("Reorg detected. Rewinding till '{0}'.", lastCommonHeader);
+
+                        lock (this.lockObject)
+                        {
+                            foreach (List<AddressBalanceChange> changes in this.addressesIndex.AddressChanges.Values)
+                                changes.RemoveAll(x => x.BalanceChangedHeight > lastCommonHeader.Height);
+                        }
+
+                        this.IndexerTip = lastCommonHeader;
+                        continue;
+                    }
+
+                    // Get next header block and process it.
+                    Block blockToProcess = this.blockStore.GetBlock(nextHeader.HashBlock);
```

Comment:
Hmm did we not say we will use CM to ask for blocks?

Reply:
Fair point, fixed